### PR TITLE
Adapt TRAP simulator spec for new digit format 

### PR DESCRIPTION
**Pull request by martenole**
- New digits are created for shared pads, so the filling of the TRAP chips doesn't need to take that into account

TODOs:
- add correct scaling factors for dy and y
- check if sorting of the digits is still needed
- MC labels not yet added
- remove MC->raw conversion part?

### CODE REVIEW

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h [hunk 1]
```diff
@@ -63,8 +63,8 @@ class Tracklet64
   //TODO convert to the actual number  regarding compliments.
   // ----- Getters for contents of tracklet word -----
   uint64_t getHCID() const { return ((mtrackletWord & hcidmask) >> hcidbs); };       // no units 0..1077
-  uint64_t getPadRow() const { return ((mtrackletWord & padrowmask) >> padrowbs); }; // in units of
-  uint64_t getColumn() const { return ((mtrackletWord & colmask) >> colbs); };       // in units of
+  uint64_t getPadRow() const { return ((mtrackletWord & padrowmask) >> padrowbs); }; // pad row number [0..15]
+  uint64_t getColumn() const { return ((mtrackletWord & colmask) >> colbs); };       // column refers to MCM position in column direction on readout board [0..3]
   uint64_t getPosition() const { return ((mtrackletWord & posmask) >> posbs); };     // in units of 0.02 pads [10bits] .. -10.22 to 10.22
```
Comment (tdietel) [inline DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: I noticed taht this comment is outdated.

File: Detectors/TRD/macros/convertRun2ToRun3Digits.C [hunk 2]
```diff
@@ -186,9 +188,19 @@ void convertRun2ToRun3Digits(TString qaOutPath = "",
...
+            int mcm = side == 0 ? (row * NMCMROBINROW + col / NCOLMCM) % NMCMROB : (row * NMCMROBINROW + (col - NCOLUMN / 2) / NCOLMCM) % NMCMROB;
+            int channel = 19 - (col % NCOLMCM);
             int tbsum = 0;
             ArrayADC adctimes;
+            bool isSharedRight = false, isSharedLeft = false;
+            if (col % NCOLMCM == 0 || col % NCOLMCM == 1) {
+              isSharedRight = true;
+            } else if (col % NCOLMCM == 17) {
+              isSharedLeft = true;
+            }
```
Comment (tdietel) [inline Detectors/TRD/macros/convertRun2ToRun3Digits.C:203]: I am not sure if I understand the code correctly, but it looks like you duplicate the entries from the run 2 digits manually. There should be a function, maybe in the AliTRDdigitsArray, that gives you access to the "hidden" ADC channels directly, without the need to duplicate.

**Review by tdietel — COMMENTED** [review]
I have looked through the PR somewhat superficially. I made one comment, but that would not stop the merge. 

Do you want to merge multiple steps, or everything at once when it's done?

File: Detectors/TRD/macros/convertRun2ToRun3Digits.C (hunk 2 above)
Comment (martenole) [reply Detectors/TRD/macros/convertRun2ToRun3Digits.C:203]: Yes, this is simply duplicating the digits manually. I can also check later on how to get the data properly. At the moment I am still struggling with the new tracklet format.

**Comment by martenole** [comment]
@tdietel I want to fix first the offset and position of the tracklets in the run 3 format. So far I could only reproduce the run 2 results using the old run 2 format. But now that I have this as a reference I can debug the new tracklet format more easily. But it is a lot fiddling with bits, signed and unsigned values and this takes some quite time. Hope that I have it done later today

File: DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h (hunk 1 above)
Comment (martenole) [reply DataFormats/Detectors/TRD/include/DataFormatsTRD/Tracklet64.h]: I updated all comments in the Tracklet64 header

File: Detectors/TRD/macros/convertRun2ToRun3Digits.C (hunk 2 above)
Comment (martenole) [reply Detectors/TRD/macros/convertRun2ToRun3Digits.C:203]: This is still not done, but I would do it in a separate PR later, I think it is not strictly required

**Comment by martenole** [comment]
Hi @bazinski and @tdietel 
from my side this PR can be merged once the checks are through. Offset and slope are fine in the run 3 tracklet format. What still needs to be done is MC labels and the charge windows also don't look right yet, but this is then for a later PR. For the global tracking this should already be something to play with.
Could you please take a look? I put the deflection correction in which was also there in Run 2 (see additional comment in the code). Not sure if it will still be done in the TRAP as well or if we should take it out.

File: Detectors/TRD/simulation/src/TrapSimulator.cxx [hunk 3]
```diff
@@ -2007,14 +1994,16 @@ void TrapSimulator::fitTracklet()
...
       temp = mult * position;
       position = temp >> 32; // take the upper 32 bits
+
+      position += yoffs;
+
       LOG(debug) << "slope = " << slope;
-      //slope = slope;//this is here as a reference for what was done in run2 //* ndrift) >> ndriftDp) + deflCorr;
-      auto oldpos = position;
+
+      slope = slope + ((deflCorr / ndrift) << ndriftDp);
```
Comment (martenole) [inline Detectors/TRD/simulation/src/TrapSimulator.cxx:2002]: This is the deflection correction I am not sure about..

Comment (bazinski) [reply Detectors/TRD/simulation/src/TrapSimulator.cxx:2002]: It can be adjusted or redone later, might be easier to diagnose further down the pipeline?

**Comment by bazinski** [comment]
@martenole the mc to raw stuff can be taken out, so the stuff about the hcid and the associated index can be removed. Turns out it was wrong in anycase. This all redone in trap2raw. I will comment in the code.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 4]
```diff
@@ -286,22 +242,23 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
    ********/
 
   auto inputDigits = pc.inputs().get<gsl::span<o2::trd::Digit>>("digitinput");
-  std::vector<o2::trd::Digit> msgDigits(inputDigits.begin(), inputDigits.end());
-  //  auto digits pc.outputs().make<std::vector<o2::trd::Digit>>(Output{"TRD", "TRKDIGITS", 0, Lifetime::Timeframe}, msgDigits.begin(), msgDigits.end());
+  std::vector<o2::trd::Digit> msgDigits(inputDigits.begin(), inputDigits.end()); // TODO: check if sorting the digits is still required. If not, this copy can be removed
```
Comment (bazinski) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: No it is no longer needs to be sorted. So long as the "double" entry digits is done in the digitizer, maybe for now check that the shared pad digits are in fact appearing twice in 2 mcm/rob pair combinations.
The sorting was to put everything into pad rows. 
However ! I assume you want tracklet data in some or other coherent sorted way, so for that purpose you might want to keep this and but change the sort criterion? Else sort tracklets pre tracking?

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 5]
```diff
@@ -314,12 +271,10 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
...
 
   /* *******
-   * reserve sizes 
+   * reserve sizes
    * *******/
-  mLinkRecords.reserve(1080 * triggerRecords.size()); // worse case scenario is all links for all events. TODO get 1080 from somewhere.
-  //TODO these must be created directly in the output as done at the top of this run method
-  //msgDigitsIndex.reserve(msgDigits.size());
-  LOG(debug) << "Read in msgDigits with size of : " << msgDigits.size() << " labels contain : " << digitMCLabels.getNElements() << " with and index size of  : " << digitMCLabels.getIndexedSize() << " and triggerrecord count of :" << triggerRecords.size();
+  mLinkRecords.reserve(2 * MAXCHAMBER * inputTriggerRecords.size()); // worse case scenario is all links for all events
```
Comment (bazinski) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: linkrecords is related to the raw data so can go. similarly all the code dependent on it.

File: Detectors/TRD/simulation/src/TrapSimulator.cxx (hunk 3 above)
Comment (martenole) [reply Detectors/TRD/simulation/src/TrapSimulator.cxx:2002]: Agree, would also leave it in for now

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 6]
```diff
@@ -330,35 +285,17 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
...
-  }
-  if (msgDigitsIndex.size() != msgDigits.size()) {
-    //error condition for sort.
-    LOG(fatal) << "Cant index digits as index and digits differ in size, this is not permitted. Digits size=" << msgDigits.size() << " and index size=" << msgDigitsIndex.size();
-  }
+  std::iota(msgDigitsIndex.begin(), msgDigitsIndex.end(), 0);
   //sort the digits array TODO refactor this intoa vector index sort and possibly generalise past merely digits.
   auto sortstart = std::chrono::high_resolution_clock::now();
-  for (auto& trig : triggerRecords) {
+  // TODO check if sorting is still needed
```
Comment (bazinski) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:215]: no its not needed in its current form. However some form of sorted maybe needed given the comment above.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 7]
```diff
@@ -376,143 +313,129 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
...
+  int currRow = msgDigits[msgDigitsIndex[0]].getRow();
+  mOldHalfChamberLinkId = LinkRecord::getHalfChamberLinkId(msgDigits[msgDigitsIndex[0]].getDetector(), msgDigits[msgDigitsIndex[0]].getROB());
+  for (const auto digitIdx : msgDigitsIndex) {
+    // while on a single padrow, populate data structures in the 8 trap simulators
+    // on change of padrow fire up trap simulators, do its thing with the data which has been written for the previous padrow
+    auto digit = &msgDigits[digitIdx];
+
     //Are we on a new half chamber ?
-    if (mOldHalfChamberLinkId != LinkRecord::getHalfChamberLinkId(detector, rob)) {
+    if (mOldHalfChamberLinkId != LinkRecord::getHalfChamberLinkId(digit->getDetector(), digit->getROB())) {
```
Comment (bazinski) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: all of this if statement can go. its logic has been moved to raw data gen.

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 8]
```diff
@@ -523,58 +446,80 @@ void TRDDPLTrapSimulatorTask::run(o2::framework::ProcessingContext& pc)
...
+    countProcessedTraps++;
+    //this one has been filled with data for the last pad row.
+    auto trapsimtimerstart = std::chrono::high_resolution_clock::now();
+    mTrapUsedCounter[trapcounter]++;
+    mTrapSimulator[trapcounter].filter();
+    mTrapSimulator[trapcounter].tracklet();
+
+    trapTracklets = mTrapSimulator[trapcounter].getTrackletArray64(); //TODO remove the copy and send the Accumulated array into the Trapsimulator
+    auto trapLabels = mTrapSimulator[trapcounter].getTrackletLabels();
+    if (!mNewTrackletHCHeaderHasBeenWritten && trapTracklets.size() != 0) { // take account of the case where we have data in the trapchip adc but no tracklets OS: Is this a TODO?
```
Comment (bazinski) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: not a todo, an explanation.

Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: but entire if and else can go. its now buried in raw generation.i.e. up to line 511

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx (hunk 7 above)
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: This also holds for lines:
554,555,556 
and the workflow can lose the rawwriter device.

**Review by bazinski — CHANGES_REQUESTED** [review]
the storage of 8 trapsimulators and the looping over 8 was only needed when the incoming pads were sorted into pad rows and then the loop was for an entire pad row. Now that digits are different this is nolonger needed but I suppose can be left in for now ? Although its not clear what defines the 8 now that its not a padrow? Or do you intend to leave the sorting digits into a padrow?

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx (hunk 4 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: I know the 8 trap chips for a specific row are no longer needed, but I wonder if it not makes sense to limit the number of TRAP chips which we keep in memory. If we fill ~65k `TrapSimulator` objects in parallel that might blow up. But I have not yet tested it.
That is why I kept the 8 (and also the digit sorting by pad row). But one could also make a loop for each chamber with 128 (96) TRAPs, maybe that is better.
Overall I have the impression that the `o2-trd-trap-sim` workflow got slower with my changes :/

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx (hunk 8 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: Well, the comment is confusing with `if trapTracklets.size() !=0` and it says to take account of the case with no tracklets. But ok, will remove the block

Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: the header still has to be added to the index even if nothing there.

Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: So should it be `trapTracklets.size() == 0` then? But anyhow, the block is removed now

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx (hunk 7 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx]: Removed the raw writer

**Comment by martenole** [comment]
@bazinski I removed all the raw stuff and rewrote the digits loop. Most of the processing time (>95%) is spent in the actual TRAP processing, so there is not much performance to be gained I think. There are 128 TRAP objects now instead of 8, so the processing is done on the level of chambers instead of pad rows.
Could you have a look again and shout if I have removed too much?

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx [hunk 9]
```diff
@@ -622,11 +330,7 @@ o2::framework::DataProcessorSpec getTRDTrapSimulatorSpec()
                            Outputs{OutputSpec{"TRD", "TRACKLETS", 0, Lifetime::Timeframe}, // this is the 64 tracklet words
                                    OutputSpec{"TRD", "TRKTRGRD", 0, Lifetime::Timeframe},
                                    /*OutputSpec{"TRD", "TRKDIGITS", 0, Lifetime::Timeframe},*/
```
Comment (bazinski) [inline Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:326]: Can you put the digits back and their associated index? Possibly an option to enable/disable it.

**Review by bazinski — DISMISSED** [review]
I think all the raw stuff is out. Chambers seem more logical way of organising it.

**Review by tdietel — DISMISSED** [review]
Looks ok as far as I can tell. Let's merge it and see.

**Review by bazinski — APPROVED** [review]

File: Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx (hunk 9 above)
Comment (martenole) [reply Detectors/TRD/workflow/src/TRDTrapSimulatorSpec.cxx:326]: Will come with a following PR

**Comment by martenole** [comment]
Hi @davidrohr,
could you please merge this when you get a chance?
Cheers,
Ole

**Pull request merged**